### ISSUE TITLE
WPA-enterprise reauthentication is not working (IDFGH-7762)

### ISSUE THREAD

**Issue by raysvmc** [espressif/esp-idf#9303]
## Environment

- Development Kit:      ESP32-DevKitC
- Kit version (for WroverKit/PicoKit/DevKitC): v4
- Module or chip used:  ESP32-WROOM-32D
- IDF version v4.4.1
- Build System:        idf.py
- Compiler version xtensa-esp32-elf esp-2021r2-patch3-8.4.0
- Operating System:     Windows
- Using an IDE?: VS Code 1.68.1
- Power Supply:         USB

## WPA enterprise authenticates reliably but never reauthenticates

We can authenticate through phase 1 and 2 and the 4 way handshake, but the reauth gets through to the very last handshake and fails.  We've collected debug logs from the esp32 and logs from the router. We've tried a variety of versions, code, and hardware, including the enterprise example code. This example is for simple test code based on the example code.

### Steps to reproduce:

1. Configure an esp32 with code to connect to an enterprise network
2. Configure the network to reauthenticate periodically
3. Allow it to connect and authenticate, and wait for the reauth.
4. Observe  sending 3/4 msg of 4-Way Handshake from the router and corresponding entries from esp-idf, near the line wifi:recv deauth, reason=0x2

We are using a Netgear WAX620 router and freeRADIUS 3.0.25 running on linux mint.

### Code to reproduce this issue

```cpp

//After this is called there is just an infinite loop along with a simple event handler function. If needed,
//I will share that too.

bool WIFI_Connect(void)
{
    esp_err_t err;

    printf("WIFI_Connect: connecting\n");

    err = esp_netif_init();

    if(err){
        printf("WIFI_Connect, esp_netif_init error %X\n", err);
        return false;
    }

    //Wifi_event_group = xEventGroupCreate();
    err = esp_event_loop_create_default();
    if(err){
        printf("WIFI_Connect, esp_event_loop_create_default error %X\n", err);
        return false;
    }

    Sta_netif = esp_netif_create_default_wifi_sta();
    if(Sta_netif == NULL){
        printf("WIFI_Connect, esp_netif_create_default_wifi_sta error\n");
        return false;
    }

    esp_netif_dhcps_stop(Sta_netif);

    esp_netif_ip_info_t info_t;
    memset(&info_t, 0, sizeof(esp_netif_ip_info_t));
    info_t.ip.addr = esp_ip4addr_aton((const char *)STATIC_IP_ADDR);
    info_t.gw.addr = esp_ip4addr_aton((const char *)GATEWAY);
    info_t.netmask.addr = esp_ip4addr_aton((const char *)NETMASK);
    esp_netif_set_ip_info(Sta_netif, &info_t);

    wifi_init_config_t cfg = WIFI_INIT_CONFIG_DEFAULT();
    err = esp_wifi_init(&cfg);
    if(err){
        printf("WIFI_Connect, esp_wifi_init error %X\n", err);
        return false;
    }

    err = esp_event_handler_instance_register(WIFI_EVENT, ESP_EVENT_ANY_ID, &_event_cb, NULL, NULL);
    if(err){
        printf("WIFI_Connect, esp_event_handler_instance_register error %X\n", err);
        return false;
    }

    err = esp_event_handler_instance_register(IP_EVENT, ESP_EVENT_ANY_ID, &_event_cb, NULL, NULL);
    if(err){
        printf("WIFI_Connect, esp_event_handler_instance_register error %X\n", err);
        return false;
    }

    err = esp_wifi_set_storage(WIFI_STORAGE_RAM);

    if(err){
        printf("WIFI_Connect, esp_wifi_set_storage error %X\n", err);
        return false;
    }


    wifi_config_t wifi_configure = {0};
    strcpy((char *)wifi_configure.sta.ssid, SSID);
    strcpy((char *)wifi_configure.sta.password, PASSWORD); 

    wifi_configure.sta.channel = 0;
    wifi_configure.sta.listen_interval = 0;
    wifi_configure.sta.scan_method = WIFI_ALL_CHANNEL_SCAN; //WIFI_ALL_CHANNEL_SCAN or WIFI_FAST_SCAN
    wifi_configure.sta.sort_method = WIFI_CONNECT_AP_BY_SIGNAL; //WIFI_CONNECT_AP_BY_SIGNAL or WIFI_CONNECT_AP_BY_SECURITY
    wifi_configure.sta.threshold.rssi = -127;
    wifi_configure.sta.pmf_cfg.capable = true;
    wifi_configure.sta.pmf_cfg.required = false;
    wifi_configure.sta.bssid_set = 0;
    memset(wifi_configure.sta.bssid, 0, 6);

    wifi_configure.sta.threshold.authmode = WIFI_AUTH_WPA2_ENTERPRISE;

    err = esp_wifi_set_mode(WIFI_MODE_STA);
    if(err){
        printf("WIFI_Connect, esp_wifi_set_mode error %X\n", err);
        return false;
    }

    err = esp_wifi_set_config(WIFI_IF_STA, &wifi_configure);
    if(err){
        printf("WIFI_Connect, esp_wifi_set_config error %X\n", err);
        return false;
    }
    
    err = esp_netif_set_hostname(Sta_netif, "somehostname");
    if(err){
        printf("WIFI_Connect, esp_netif_set_hostname error %X\n", err);
        return false;
    }

    err = esp_wifi_sta_wpa2_ent_set_identity((uint8_t *)IDENT, strlen(IDENT));
    if(err){
            printf("WIFI_Connect, esp_wifi_sta_wpa2_ent_set_identity %s, error %X\n", IDENT, err);
            return false;
    }

    err = esp_wifi_sta_wpa2_ent_set_username((uint8_t *)IDENT, strlen(IDENT));
    if(err){
            printf("WIFI_Connect, esp_wifi_sta_wpa2_ent_set_username %s, error %X\n", IDENT, err);
            return false;
    }

    err = esp_wifi_sta_wpa2_ent_set_password((uint8_t *)PASSWORD, strlen(PASSWORD));
    if(err){
            printf("WIFI_Connect, esp_wifi_sta_wpa2_ent_set_password error %X\n", err);
            return false;
    }

    err = esp_wifi_sta_wpa2_ent_enable();
    if(err){
            printf("WIFI_Connect, esp_wifi_sta_wpa2_ent_enable error %X\n", err);
            return false;
    }

    err = esp_wifi_start();
    if(err){
        printf("WIFI_Connect, esp_wifi_start error %X\n", err);
        return false;
    }   

    printf("WIFI_Connect: should connect now.\r\n");

    return true;
}
```

## Debug Logs

```
Router:

Jun  1 09:46:48 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c IEEE 802.11: authenticated
Jun  1 09:46:48 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c IEEE 802.11: associated (aid 1)
Jun  1 09:46:49 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c WPA: sending 1/4 msg of 4-Way Handshake
Jun  1 09:46:49 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c WPA: received EAPOL-Key frame (2/4 Pairwise)
Jun  1 09:46:49 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c WPA: sending 3/4 msg of 4-Way Handshake
Jun  1 09:46:49 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c WPA: received EAPOL-Key frame (4/4 Pairwise)
Jun  1 09:46:49 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c WPA: pairwise key handshake completed (RSN)
Jun  1 09:46:49 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c IEEE 802.1X: authenticated - EAP type: 25 (PEAP)
Jun  1 09:46:49 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c WPA: Added PMKSA cache entry (IEEE 802.1X)
Jun  1 09:46:49 nddmp[5395]: alarm : seqNo-[22480], level-[INFO], info-[10-97-BD-D4-A1-8C wireless client connected]
Jun  1 09:46:49 nddmp[5395]: alarm : seqNo-[22481], level-[OTHER], info-[{"ip":"-","mac":"10-97-BD-D4-A1-8C","hname":"10-97-BD-D4-A1-8C","txR":"1 Mbps","rxR":"72.20 Mbps","ssid":"VMC-ENTERPRISE-1","dOs":"Unknown Device OS","dType":"Unknown Device Type","mode":"11NG","status":"Authenticated","bssid":"6C-CD-D6-5F-A9-C1","chnl":"6","chWidth":"20 MHz","rssi":"64","state":"QOS/ERP/HT","type":"wpa2","idle":"0","time":"00:00:00","txB":"350","rxB":"3632","pmf":"none","vlanID":"1","Username":"TBD5","radio":"0"}]
Jun  1 09:47:06 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:47:06 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:47:36 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:47:36 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:48:06 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:48:06 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:48:24 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:48:24 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:48:36 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:48:36 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:48:52 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c WPA: sending 1/2 msg of Group Key Handshake
Jun  1 09:48:52 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c WPA: received EAPOL-Key frame (2/2 Group)
Jun  1 09:49:06 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:49:06 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:49:36 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:49:36 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:50:06 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:50:06 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:50:36 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:50:36 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:51:06 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:51:06 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:51:36 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:51:36 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:51:49 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c IEEE 802.1X: authenticated - EAP type: 25 (PEAP)
Jun  1 09:51:49 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c WPA: Added PMKSA cache entry (IEEE 802.1X)
Jun  1 09:51:49 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c WPA: sending 1/4 msg of 4-Way Handshake
Jun  1 09:51:49 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c WPA: received EAPOL-Key frame (2/4 Pairwise)
Jun  1 09:51:49 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c WPA: sending 3/4 msg of 4-Way Handshake
Jun  1 09:51:50 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c WPA: sending 3/4 msg of 4-Way Handshake
Jun  1 09:51:51 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c WPA: sending 3/4 msg of 4-Way Handshake
Jun  1 09:51:52 hostapd: wifi0vap0: STA 10:97:bd:d4:a1:8c IEEE 802.11: disconnected
Jun  1 09:51:52 nddmp[5395]: alarm : seqNo-[22482], level-[INFO], info-[10-97-BD-D4-A1-8C wireless client could not obtain IP address]
Jun  1 09:51:52 nddmp[5395]: alarm : seqNo-[22483], level-[INFO], info-[10-97-BD-D4-A1-8C wireless client disconnected]
Jun  1 09:52:06 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1
Jun  1 09:52:06 configd[5330]: get_cur_channel: None of the interfaces are in UP state for wifi1


Serial Port:
(time zones don't match)

"6/1/2022 11:46:47 AM",D (8615) esp_netif_lwip: starting dhcp client[0m
"6/1/2022 11:46:47 AM",D (8625) esp_netif_lwip: call api in lwip: ret=0x0, give sem[0m
"6/1/2022 11:46:47 AM",D (8625) event: running post WIFI_EVENT:4 with handler 0x400d67f4 and context 0x3ffc4e74 on loop 0x3ffbcc54[0m
"6/1/2022 11:46:47 AM",[wifi] Event ................: STA_CONNECTED
"6/1/2022 11:48:50 AM",W (131655) wifi:<ba-add>idx:0 (ifx:0, 6c:cd:d6:5f:a9:c1), tid:6, ssn:15, winSize:64
"6/1/2022 11:48:50 AM",D (131665) wifi:obss scan is disabled
"6/1/2022 11:48:50 AM",D (131665) wifi:start obss scan: obss scan is stopped
"6/1/2022 11:51:47 AM",D (308825) wifi:obss scan is disabled
"6/1/2022 11:51:47 AM",D (308825) wifi:start obss scan: obss scan is stopped
"6/1/2022 11:51:50 AM",D (312085) wifi:recv deauth, reason=0x2
"6/1/2022 11:51:50 AM",I (312085) wifi:state: run -> init (2c0)
"6/1/2022 11:51:50 AM",D (312085) wifi:recv deauth/disassoc, stop beacon/connect timer
"6/1/2022 11:51:50 AM",D (312085) wifi:connect status 5 -> 4
"6/1/2022 11:51:50 AM",D (312095) wifi:sta leave
"6/1/2022 11:51:50 AM",I (312095) wifi:pm stop, total sleep time: 288383634 us / 303586331 us

"6/1/2022 11:51:50 AM",D (312095) wifi:stop CSA timer
"6/1/2022 11:51:50 AM",W (312105) wifi:<ba-del>idx
"6/1/2022 11:51:50 AM",D (312105) wifi:remove 6c:cd:d6:5f:a9:c1 from rc list
"6/1/2022 11:51:50 AM",I (312105) wifi:new:<6,0>, old:<6,0>, ap:<255,255>, sta:<6,0>, prof:1
"6/1/2022 11:51:50 AM",D (312115) wifi:filter: set rx policy=8
"6/1/2022 11:51:50 AM",D (312115) wifi:Send disconnect event, reason=2, AP number=0
"6/1/2022 11:51:50 AM",D (312125) event: running post WIFI_EVENT:5 with handler 0x400e89d0 and context 0x3ffbde88 on loop 0x3ffbcc54[0m
"6/1/2022 11:51:50 AM",D (312135) esp_netif_handlers: esp_netif action disconnected with netif0x3ffbdb80 from event_id=5[0m
"6/1/2022 11:51:50 AM",D (312145) esp_netif_lwip: check: remote, if=0x3ffbdb80 fn=0x400dd494
"6/1/2022 11:51:50 AM",[0m
"6/1/2022 11:51:50 AM",D (312155) esp_netif_lwip: esp_netif_down_api esp_netif:0x3ffbdb80[0m
"6/1/2022 11:51:50 AM",D (312155) esp_netif_lwip: esp_netif_start_ip_lost_timer esp_netif:0x3ffbdb80[0m
"6/1/2022 11:51:50 AM",D (312165) esp_netif_lwip: if0x3ffbdb80 start ip lost tmr: no need start because netif=0x3ffbdc00 interval=120 ip=0[0m
"6/1/2022 11:51:50 AM",D (312175) esp_netif_lwip: check: local, if=0x3ffbdb80 fn=0x400ddc6c
"6/1/2022 11:51:50 AM",[0m
"6/1/2022 11:51:50 AM",D (312185) esp_netif_lwip: esp_netif_update_default_netif_lwip 0x3ffbdb80[0m
"6/1/2022 11:51:50 AM",V (312185) esp_netif_objects: esp_netif_next_unsafe 0x0[0m
"6/1/2022 11:51:50 AM",V (312195) esp_netif_lwip: esp_netif_is_netif_up esp_netif:0x3ffbdb80[0m
"6/1/2022 11:51:50 AM",V (312195) esp_netif_objects: esp_netif_next_unsafe 0x3ffbdb80[0m
"6/1/2022 11:51:50 AM",D (312205) esp_netif_lwip: call api in lwip: ret=0x0, give sem[0m
"6/1/2022 11:51:50 AM",D (312215) event: running post WIFI_EVENT:5 with handler 0x400d67f4 and context 0x3ffc4e74 on loop 0x3ffbcc54[0m
"6/1/2022 11:51:50 AM",[wifi] Event ................: STA_DISCONNECTED

```

**Comment by hpdhabe** [comment]
Hi @raysvmc,
Can you please attach the debug logs and a sniffer capture to this issue if possible? It will be really helpful in debugging the issue.
Please make sure the flag `WPA_DEBUG_PRINT` is enabled before capturing the debug logs. This can be found under the `Supplicant` component in `Component Config` menu of `menuconfig`.
If you need more help with capturing 82011 packets, refer to the following link: https://github.com/espressif/esp-idf/issues/8687#issuecomment-1105070843

**Comment by raysvmc** [comment]
Thank you @hpdhabe for looking into this. We're trying to get set back up to make another test but it might take a week or so for various reasons I won't bore you with. I have set WPA_DEBUG_PRINT and configured a laptop for monitor mode sniffing.

**Comment by AxelLin** [comment]
@raysvmc  Any update regarding this issue?

**Comment by raysvmc** [comment]
Apologies for the delay @hpdhabe and @AxelLin , we've had a series of other problems in the way, but it seems like we're almost ready to make another test. I will push for that.

**Comment by AxelLin** [comment]
@raysvmc 
Just wondering if this is still an issue or not.
If you still hit the issue, please provide debug logs and a sniffer capture. see https://github.com/espressif/esp-idf/issues/9303#issuecomment-1181643113

**Comment by raysvmc** [comment]
@AxelLin this is very much an issue for us and our customers. I'm unable to perform the test myself as I am remote and don't have the system setup here or the knowledge to set it up. Two other engineers have been performing the tests, and I have been asking repeatedly that it get done ASAP. They have been working on it, but between other emergencies, vacations, and trouble with the router configuration, it's not quite done. I will continue pushing for it and I apologize for the delay. I'm hoping this week!

**Comment by raysvmc** [comment]
We have a test finally @hpdhabe @AxelLin and I will share:
capture5-9.9.2022.1555.txt from the debug port
messages.txt from the router
The packet capture .pcap is not being allowed - I will investigate.
The behavior is a bit different but it's failing the same way as far as I can tell.
[capture5-9.9.2022.1555.txt](https://github.com/espressif/esp-idf/files/9539212/capture5-9.9.2022.1555.txt)
[messages.txt](https://github.com/espressif/esp-idf/files/9539213/messages.txt)

**Comment by raysvmc** [comment]
[(Acrylic) VMC-ENTERPRISE-1_9.9.2022.1555.csv](https://github.com/espressif/esp-idf/files/9539220/Acrylic.VMC-ENTERPRISE-1_9.9.2022.1555.csv)

**Comment by raysvmc** [comment]
[WAX620-NETGEAR-VMC br-lan-2020-03-03__164958-capture.csv](https://github.com/espressif/esp-idf/files/9539224/WAX620-NETGEAR-VMC.br-lan-2020-03-03__164958-capture.csv)
This is a capture of the router's network. I don't know if this is at all helpful.

**Comment by Shreyas0-7** [comment]
Hey @raysvmc 
Can you please retry with this patch 
[wpa_reauth.zip](https://github.com/espressif/esp-idf/files/9625282/wpa_reauth.zip)

**Comment by raysvmc** [comment]
Shreyas0-7, thanks for the suggestion, we tried that and had about the same results I'm afraid.
[AuthIssue092222.zip](https://github.com/espressif/esp-idf/files/9629657/AuthIssue092222.zip)

**Comment by Shreyas0-7** [comment]
Hey @raysvmc is it possible for you to log hostapd logs as well?
You can do it by passing arguments -dd and -K while starting hostapd.

**Comment by raysvmc** [comment]
It looks like we have that log. Is this it?
[hostapd.txt](https://github.com/espressif/esp-idf/files/9634894/hostapd.txt)

**Comment by Shreyas0-7** [comment]
Hey @raysvmc thanks for your patience and your co-ordination, can you please retry with this patch.
[wpa2_reauth.zip](https://github.com/espressif/esp-idf/files/9668973/wpa2_reauth.zip)

**Comment by raysvmc** [comment]
Thanks @Shreyas0-7, I will try to compile that today and it might be tested tomorrow.

**Comment by se-k32** [comment]
Hi, I am joining this conversation because I think we have the same problem (apologize if not; in that case, I will create a new post). One of our clients has a problem that all ESP32 devices gets disconnected precisely 3 hours after being connected to AP. AP is using WPA2-PSK authentication (different as used by @raysvmc). The disconnect reason is WIFI_REASON_AUTH_EXPIRE (2) or sometimes WIFI_REASON_NOT_AUTHED (6). After that, devices reconnect to AP and the same thing happens again after 180 minutes. We think that the AP requests some kind of re-authentication, but ESP fails to respond to it, so AP kicks him off. We are using esp-idf 4.4.3. We tried the patch from wpa2_reauth.zip, but sadly it did not help. We don't have precise logs, but we are happy to help testing it.

**Comment by AxelLin** [comment]
@se-k32 
Could you create a new issue?
Please provide sniffer log and device verbose log if possible, and what is the AP brand/model?
BTW, esp-idf-v4.4.3 is not yet available, please provide correct version (git describe --tags).

**Comment by se-k32** [comment]
Yes no problem, I have created it (https://github.com/espressif/esp-idf/issues/9896#issue-1393698360). I thought that the solution/problem is the same as this one. Also sorry for the wrong esp-idf version, it should be 4.4.2.

**Comment by Shreyas0-7** [comment]
Thanks for retry @raysvmc. By applying this patch we were able to fix this bug in our test environment but as this was also seen again in [9896](https://github.com/espressif/esp-idf/issues/9896). To investigate it furthermore we will require sniffer capture on the particular channel on which AP is turned on. Also in sniffer capture it is observed that only beacon frames are seen. Please make sure that you are able to capture the handshake also. Thank you.

**Comment by raysvmc** [comment]
Thanks @Shreyas0-7 and I apologize for not checking the sniffer capture. We're working on that.

**Comment by AxelLin** [comment]
> Thanks @Shreyas0-7 and I apologize for not checking the sniffer capture. We're working on that.

Any update regarding the sniffer log? @raysvmc

**Comment by raysvmc** [comment]
@AxelLin unfortunately not yet. There has been some kind of difficulty in making this work and I will see if I can help. Thanks for asking!

**Comment by raysvmc** [comment]
@Shreyas0-7 @AxelLin this last test seems to be working now, but I will share what we got on this one in case it's of interest. We'll have to see if this is repeatable. There are two tests, each of which reauths successfully with a few handshake resends. I summarized them in NotesFromSerialCapture.txt. Thanks!
[AuthIssue102722.zip](https://github.com/espressif/esp-idf/files/9883240/AuthIssue102722.zip)

**Comment by raysvmc** [comment]
@Shreyas0-7 @AxelLin we've run two tests, and the second one was fine. I'm sharing the first one which worked sometimes but had different problems authenticating. There are two reboots so it's kind of three tests in one and I wrote up a text file cross-referencing the logs. I hope you can follow it, as I had a difficult time. There's a LAN capture but I didn't share it - please let me know if that would help.
[AuthIssue102722a.zip](https://github.com/espressif/esp-idf/files/9891040/AuthIssue102722a.zip)

**Comment by AxelLin** [comment]
@Shreyas0-7 Any update on this issue? Did you check above logs?

**Comment by Shreyas0-7** [comment]
@raysvmc @AxelLin As seen in logs and sniffer capture AP sends EAPOL message 3 of 4 twice with different replay counters, we reply to the first one encrypting with older keys but AP expects 4 of 4 for the latest replay counter to which we reply encrypting using new keys. This situation cannot be handled at this moment, also you can observe the same behavior in upstream.

**Comment by AxelLin** [comment]
> Thanks for retry @raysvmc. By applying this patch we were able to fix this bug in our test environment but as this was also seen again in [9896](https://github.com/espressif/esp-idf/issues/9896). 

@Shreyas0-7 
https://github.com/espressif/esp-idf/issues/9896#issuecomment-1272018890 set esp_wifi_set_inactive_time(WIFI_IF_STA, 10); to avoid the issue.
Is that a reasonable change?  Any bad side-effect with setting inactive time to 10 sec?
Or is that actually a different issue?

**Comment by Shreyas0-7** [comment]
@AxelLin Issue you mentioned is completely different and setting esp_wifi_set_inactive_time(WIFI_IF_STA, 10); will have no effect here. 
@raysvmc Yes AP should have waited longer to send message 3 of 4 again. By upstream behavior, I meant wpa_supplicant which is used by Linux.

**Comment by AxelLin** [comment]
@Shreyas0-7 

Then is this https://github.com/espressif/esp-idf/issues/9303#issuecomment-1261472280 a vaild fix? (since you already confirm it fixed the bug in your test environment.)
The fix is not yet available in any branches, so it's not clear to me if you need more testing or it's not a good fix?

**Comment by Shreyas0-7** [comment]
@AxelLin The patch I provided contained the fix. I have raised the corresponding MR and it's under testing, it will be available once it gets merged.

**Comment by raysvmc** [comment]
@Shreyas0-7 these might be dumb questions, but, first, I applied both patches wpa_reauth and wpa2_reauth; should I have only applied one at a time?
Second, I understand that the AP is not waiting long enough, but is there another problem with the wpa_supplicant used by Linux, and is this the RADIUS server? Thanks for your patience!

**Comment by AxelLin** [comment]
@Shreyas0-7 
Fix for v4.4:  24efd4a505c27d97c8bccfa9e8c7a6258536b852 
I think v4.3 also need the fix.

**Comment by kapilkedawat** [comment]
> @Shreyas0-7 these might be dumb questions, but, first, I applied both patches wpa_reauth and wpa2_reauth; should I have only applied one at a time? 

Yes, Please use final fix (https://github.com/espressif/esp-idf/commit/24efd4a505c27d97c8bccfa9e8c7a6258536b852)

> Second, I understand that the AP is not waiting long enough, but is there another problem with the wpa_supplicant used by Linux, and is this the RADIUS server? Thanks for your patience!
What we mean is, same issue is present in upstream linux based distributions as well. the solution will be to keep both new key and old key in hardware till device confirms that the new key will be used for further connection which cannot be supported.

> I think v4.3 also need the fix.
Yes, the fix has been backported and will be available shortly.

**Comment by raysvmc** [comment]
@kapilkedawat Thank you!

**Comment by AxelLin** [comment]
@kapilkedawat @Shreyas0-7 
Looks like v5.0 tag does not include this fix?

**Comment by Shreyas0-7** [comment]
The fix is backported to v5.0 and is merged recently
